superduperdb/ext/llm/model.py

```diff
@@ -0,0 +1,459 @@
+
+        lora_bias (`str`, *optional*, defaults to "none"):
+            Lora bias.
+
+        max_length (`int`, *optional*, defaults to 512):
+            Maximum source sequence length during training.
+
+    """
+
+    lora_r: int = 8
+    lora_alpha: int = 16
+    lora_dropout: float = 0.05
+    lora_target_modules: t.Optional[t.List[str]] = None
+    lora_weight_path: str = ""
+    lora_bias: str = "none"
+    max_length: t.Optional[int] = 512
+
+    __doc__ = __doc__.format(training_arguments_doc=TrainingArguments.__doc__)
+
+
+@functools.wraps(LLMTrainingArguments)
+def LLMTrainingConfiguration(identifier: str, **kwargs) -> _TrainingConfiguration:
+    return _TrainingConfiguration(identifier=identifier, kwargs=kwargs)
+
+
+@dc.dataclass
+class LLM(Model):
+    """
+    LLM model based on `transformers` library.
+    Parameters:
+    : param identifier: model identifier
+    : param model_name_or_path: model name or path
+    : param bits: quantization bits, [4, 8], default is None
+    : param model_kwargs: model kwargs,
+        all the kwargs will pass to `transformers.AutoModelForCausalLM.from_pretrained`
+    : param tokenizer_kwags: tokenizer kwargs,
+        all the kwargs will pass to `transformers.AutoTokenizer.from_pretrained`
+    : param prompt_template: prompt template, default is "{input}"
+    : param prompt_func: prompt function, default is None
+    """
+
+    identifier: str = ""
+    model_name_or_path: str = "facebook/opt-125m"
+    bits: t.Optional[int] = None
+    object: t.Optional[transformers.Trainer] = None
+    model_kwargs: t.Dict = dc.field(default_factory=dict)
+    tokenizer_kwags: t.Dict = dc.field(default_factory=dict)
+    prompt_template: str = "{input}"
+    prompt_func: t.Optional[t.Callable] = dc.field(default=None)
+
+    def __post_init__(self):
+        self.identifier = self.identifier or self.model_name_or_path
+        # overwrite model kwargs
+        if self.bits is not None:
+            self.model_kwargs["load_in_4bit"] = self.bits == 4
+            self.model_kwargs["load_in_8bit"] = self.bits == 8
+        super().__post_init__()
+
+    def init_model_and_tokenizer(self):
+        model = AutoModelForCausalLM.from_pretrained(
+            self.model_name_or_path,
+            **self.model_kwargs,
+        )
+
+        tokenizer = AutoTokenizer.from_pretrained(
+            self.model_name_or_path,
+            **self.tokenizer_kwags,
+        )
+        tokenizer.pad_token = tokenizer.pad_token or tokenizer.unk_token
+        return model, tokenizer
+
+    def create_trainer(
+        self, train_dataset, eval_dataset, training_args, **kwargs
+    ) -> transformers.Trainer:
+        trainer = transformers.Trainer(
+            model=self.model,
+            tokenizer=self.tokenizer,
+            args=training_args,
+            train_dataset=train_dataset,
+            eval_dataset=eval_dataset,
+            **kwargs,
+        )
+        return trainer
+
+    def init(self):
```

Comment:
is it used?

Reply:
Yes, it will call by `ensure_initialized`

Then, we can lazy load the model until it is actually used

```python
    @ensure_initialized
    def _predict(
        self,
        X: t.Union[str, t.List[str], t.List[dict[str, str]]],
        one: bool = False,
        **kwargs: t.Any,
    ):

```